examples/multi_agent/tic_tac_toe.py

```diff
@@ -0,0 +1,197 @@
+import os
+import pprint
+from torch.utils.tensorboard import SummaryWriter
+import torch
+import argparse
+import numpy as np
+from typing import Tuple, Optional
+
+from tianshou.env import MAEnv, VectorEnv
+from tianshou.policy import MADQNPolicy
+from tianshou.data import Collector, ReplayBuffer
+from tianshou.trainer import offpolicy_trainer
+
+from MANet import Net
+
+
+class TicTacToeEnv(MAEnv):
+    board = [[0, 0, 0],
+             [0, 0, 0],
+             [0, 0, 0]]
+
+    def __init__(self):
+        super().__init__(None)
+        self.current_board = np.array(TicTacToeEnv.board)
+        self.current_player = 0
+
+    def reset(self) -> dict:
+        self.current_board = np.array(TicTacToeEnv.board)
+        self.current_player = 0
+        return {
+            'agent_id': 0,
+            'obs': np.array(self.current_board),
+            'legal_actions': set(i for i in range(9))
+        }
+
+    def step(self, action: np.ndarray
+             ) -> Tuple[dict, np.ndarray, np.ndarray, dict]:
+        action = action.item(0)
+        assert 0 <= action < 9
+        assert self.current_board.item(action) == 0
+        self._move(action)
+        legal_actions = {i for i, b in enumerate(
+            self.current_board.flatten().tolist()) if b == 0}
+        is_win, is_opponent_win = False, False
+        is_win = self._test_win()
+        # the game is over when one wins or there is only one empty place
+        done = is_win
+        if len(legal_actions) == 1:
+            done = True
+            self._move(list(legal_actions)[0])
+            is_opponent_win = self._test_win()
+        if is_win:
+            reward = 1
+        elif is_opponent_win:
+            reward = -1
+        else:
+            reward = 0
+        obs = {
+            'agent_id': self.current_player,
+            'obs': np.array(self.current_board),
+            'legal_actions': legal_actions
+        }
+        return obs, np.array(reward), np.array(done), {}
```

Comment:
it should be me necessary for `done` or `reward` to be a `np.ndrray`, but as you wish.

Reply:
I just add this to pass type check.

Reply:
Can you clarify which type check please ?

Reply:
The function signature of ``step()``.

Reply:
OK I see. This is problematic, since it goes against the specification of OpenAI Gym:
https://github.com/openai/gym/blob/345c65973fc7160d8be374745a60c36869d8accc/gym/core.py#L52

Reply:
> OK I see. This is problematic, since it goes against the specification of OpenAI Gym:
> https://github.com/openai/gym/blob/345c65973fc7160d8be374745a60c36869d8accc/gym/core.py#L52

I do not think this is problematic. In the future, we may support multi-agent environment whose reward is a vector (with each item the reward of each player).

Reply:
It is in a sense that you require something that it is not part of OpenAI Gym, so Tianshou RL algorithm will raise exception on any environment implementing the specification of OpenAI Gym in one day typing becomes strict and (almost) not only cosmetic as it is the case for now.

Reply:
> It is, in a sense that you require something that it is not part of OpenAI Gym, so Tianshou RL algorithm will raise exception on any environment implementing the specification of OpenAI Gym in one day typing becomes strict and (almost) not only cosmetic as it is the case for now.

So you mean we must stay in the scope of OpenAI Gym? That seems too restricted, as multi-agent RL is apparently not in the scope of OpenAI Gym.

Reply:
Of course not, why to be so restrictive ? 

No I'm saying that Tianshou should support `Union[np.ndarray, float]` for reward and `Union[np.ndarray, bool]` for done.